onnxruntime/contrib_ops/cpu/bert/multihead_attention.cc

```diff
@@ -0,0 +1,464 @@
+                                                                      nullptr,
+                                                                      &parameters,
+                                                                      num_heads_,
+                                                                      scale,
+                                                                      mask_filter_value_,
+                                                                      past_present_share_buffer));
+
+  const int batch_size = parameters.batch_size;
+  const int q_sequence_length = parameters.sequence_length;
+  const int kv_sequence_length = parameters.kv_sequence_length;
+  const int past_kv_sequence_length = parameters.past_sequence_length;
+  const int total_kv_sequence_length = parameters.total_sequence_length;
+  int qk_head_size = parameters.head_size;
+  int v_head_size = parameters.v_head_size;
+  int qk_hidden_size = parameters.hidden_size;
+  int v_hidden_size = parameters.v_hidden_size;
+
+  std::vector<int64_t> output_shape(3);
+  output_shape[0] = static_cast<int64_t>(batch_size);
+  output_shape[1] = static_cast<int64_t>(q_sequence_length);
+  output_shape[2] = static_cast<int64_t>(parameters.v_hidden_size);
+  Tensor* output = context->Output(0, output_shape);
+
+  auto element_type = DataTypeImpl::GetType<T>();
+  const auto* qkv_bias = (bias == nullptr) ? nullptr : bias->Data<T>();
+  const int q_bias_offset = 0;
+  const int k_bias_offset = qk_hidden_size;
+  const int v_bias_offset = 2 * qk_hidden_size;
+
+  const int num_q_elements = batch_size * num_heads_ * q_sequence_length * qk_head_size;
+  const int num_k_elements = batch_size * num_heads_ * kv_sequence_length * qk_head_size;
+  const int num_v_elements = batch_size * num_heads_ * kv_sequence_length * v_head_size;
+
+  bool kv_BNSH = key != nullptr && value != nullptr && key->Shape().GetDims().size() == 4 && value->Shape().GetDims().size() == 4;
+  bool packed_qkv = query->Shape().GetDims().size() == 3 && key == nullptr && value == nullptr;
+  
+  // If optional outputs aren't needed, present_k and present_v will be null
+  std::vector<int64_t> present_k_shape({static_cast<int64_t>(batch_size), static_cast<int64_t>(num_heads_), static_cast<int64_t>(total_kv_sequence_length), static_cast<int64_t>(qk_head_size)});
+  std::vector<int64_t> present_v_shape({static_cast<int64_t>(batch_size), static_cast<int64_t>(num_heads_), static_cast<int64_t>(total_kv_sequence_length), static_cast<int64_t>(v_head_size)});
+  Tensor* present_k = context->Output(1, present_k_shape);
+  Tensor* present_v = context->Output(2, present_v_shape);
+
+  bool separate_present_kv = present_k != nullptr && present_v != nullptr;
+
+  Tensor* past_kv = nullptr;
+  OrtValue past;
+  if (past_key != nullptr && past_value != nullptr) {
+    // Concatenate past_key and past_value into past_kv
+    // We assume qk_head_size == v_head_size
+    AllocatorPtr allocator;
```

Comment:
it's better to create allocator and buffer once and initialize OrtValue with buffer ptr.